crates/sui-core/src/narwhal_manager/mod.rs

```diff
@@ -102,37 +102,69 @@ impl NarwhalManager {
         tracing::info!("Starting up Narwhal for epoch {}", committee.epoch());
 
         // start primary
-        self.primary_node
-            .start(
-                self.primary_keypair.copy(),
-                self.network_keypair.copy(),
-                Arc::new(ArcSwap::new(committee.clone())),
-                shared_worker_cache.clone(),
-                &store,
-                execution_state,
-            )
-            .await
-            .expect("Unable to start Narwhal Primary");
+        let mut retries_left = 2;
+        while retries_left > 0 {
+            match self
+                .primary_node
+                .start(
+                    self.primary_keypair.copy(),
+                    self.network_keypair.copy(),
+                    Arc::new(ArcSwap::new(committee.clone())),
+                    shared_worker_cache.clone(),
+                    &store,
+                    execution_state.clone(),
+                )
+                .await
+            {
+                Ok(_) => {
+                    break;
+                }
+                Err(e) => {
+                    if retries_left <= 0 {
+                        panic!("Unable to start Narwhal Primary: {:?}", e);
+                    }
+                    tracing::error!("Unable to start Narwhal Primary: {:?}, retrying", e);
+                    retries_left -= 1;
+                    std::thread::sleep(std::time::Duration::from_millis(100));
```

Comment:
This should probably be `tokio::time::sleep` , otherwise we'll make the whole thread sleep instead.

Reply:
Also I would advise giving a bit more time here - even 1 second